carapace-server/src/main/java/org/carapaceproxy/server/backends/BackendHealthStatus.java

```diff
@@ -19,75 +19,122 @@
-        this.reportedAsUnreachable = reportedAsUnreachable;
+    public long getLastUnreachableTs() {
+        return lastUnreachableTs;
     }
 
-    public long getReportedAsUnreachableTs() {
-        return reportedAsUnreachableTs;
+    void reportAsUnreachable(final long timestamp, final String cause) {
+        LOG.info("{}: reportAsUnreachable {} cause {}", hostPort, new Timestamp(timestamp), cause);
+        this.lastUnreachableTs = timestamp;
+        this.status = Status.DOWN;
+        this.connections.set(0);
     }
 
-    public void setReportedAsUnreachableTs(long reportedAsUnreachableTs) {
-        this.reportedAsUnreachableTs = reportedAsUnreachableTs;
+    void reportAsReachable(final long timestamp) {
+        this.lastReachableTs = timestamp;
+        if (this.lastReachableTs - this.lastUnreachableTs >= WARMUP_MILLIS) {
+            this.status = Status.WARM;
+        } else {
+            this.status = Status.COLD;
+        }
     }
 
-    void reportAsUnreachable(long timestamp) {
-        LOG.log(Level.INFO, "{0}: reportAsUnreachable {1}", new Object[]{hostPort, new java.sql.Timestamp(timestamp)});
-        reportedAsUnreachableTs = timestamp;
-        reportedAsUnreachable = true;
+    @Override
+    public String toString() {
+        return "BackendHealthStatus{"
+                + " hostPort=" + this.hostPort
+                + ", status=" + this.status
+                + ", lastUnreachableTs=" + this.lastUnreachableTs
+                + ", lastReachableTs=" + this.lastReachableTs
+                + ", lastProbe=" + this.lastProbe
+                + '}';
     }
 
-    void reportAsReachable() {
-        reportedAsUnreachable = false;
-        reportedAsUnreachableTs = 0;
+    public int getConnections() {
+        return this.connections.get();
     }
 
-    public boolean isAvailable() {
-        return !reportedAsUnreachable;
+    public int incrementConnections() {
+        return this.connections.incrementAndGet();
     }
 
-    @Override
-    public String toString() {
-        return "BackendHealthStatus{" + "hostPort=" + hostPort + ", reportedAsUnreachable=" + reportedAsUnreachable + ", reportedAsUnreachableTs=" + reportedAsUnreachableTs + '}';
+    public int decrementConnections() {
+        if (this.connections.getAcquire() > 0) {
+            return this.connections.decrementAndGet();
+        }
+        return 0;
     }
 
+    /**
+     * The enum models a simple status of the backend.
+     */
+    public enum Status {
+        /**
+         * The backend is unreachable.
+         */
+        DOWN,
+        /**
+         * The backend is reachable, but not since long.
+         * When in this safe, it is reasonable to assume that it is still warming-up,
+         * so it would be a sensible decision not to overload it with requests.
+         */
+        COLD,
+        /**
+         * The backend is reachable and was so since a reasonably-long time.
+         */
+        WARM
```

Comment:
If the backend is relatively stable for a sufficiently long time, perhaps we could use something like:
ACTIVE, HEALTHY, or STABLE.
In this case, STABLE might sound better as the focus is on stability over time.

What do you think @NiccoMlt  ?

Reply:
I agree, I was out of ideas on which name to use here ahah

Reply:
So we want to call them:

- `DOWN`
- `COLD`
- `STABLE`

Right?

Reply:
Yes, can go for me